src/elfpy/markets.py

```diff
@@ -51,83 +106,154 @@ def __init__(
+        )
+        agent_action.print_description_string()
+
         # for each position, specify how to forumulate trade and then execute
-        if user_action.action_type == "open_long":  # buy to open long
-            user_action.direction = "out"  # calcOutGivenIn
-            user_action.token_out = "pt"  # buy unknown PT with known base
-            market_deltas, wallet_deltas = self._open_long(user_action)
-        elif user_action.action_type == "close_long":  # sell to close long
-            user_action.direction = "out"  # calcOutGivenIn
-            user_action.token_out = "base"  # sell known PT for unkonwn base
-            market_deltas, wallet_deltas = self._close_long(user_action)
-        elif user_action.action_type == "open_short":  # sell PT to open short
-            user_action.direction = "out"  # calcOutGivenIn
-            user_action.token_out = "base"  # sell known PT for unknown base
-            market_deltas, wallet_deltas = self._open_short(user_action)
-        elif user_action.action_type == "close_short":  # buy PT to close short
-            user_action.direction = "in"  # calcInGivenOut
-            user_action.token_in = "base"  # buy known PT for unknown base
-            market_deltas, wallet_deltas = self._close_short(user_action)
+        if agent_action.action_type == "open_long":  # buy to open long
+            agent_action.direction = "out"  # calcOutGivenIn
+            agent_action.token_out = "pt"  # sell known base for unknown PT
+            market_deltas, agent_deltas = self._open_long(agent_action)
+        elif agent_action.action_type == "close_long":  # sell to close long
+            agent_action.direction = "out"  # calcOutGivenIn
+            agent_action.token_out = "base"  # sell known PT for unknown base
+            market_deltas, agent_deltas = self._close_long(agent_action)
+        elif agent_action.action_type == "open_short":  # sell PT to open short
+            agent_action.direction = "in"  # calcOutGivenIn
+            agent_action.token_out = "pt"  # sell known PT for unknown base
+            market_deltas, agent_deltas = self._open_short(agent_action)
+        elif agent_action.action_type == "close_short":  # buy PT to close short
+            agent_action.direction = "in"  # calcOutGivenIn
+            agent_action.token_in = "pt"  # buy known PT for unknown base
+            market_deltas, agent_deltas = self._close_short(agent_action)
+        elif agent_action.action_type == "add_liquidity":
+            market_deltas, agent_deltas = self._add_liquidity(agent_action)
+        elif agent_action.action_type == "remove_liquidity":
+            market_deltas, agent_deltas = self._remove_liquidity(agent_action)
         else:
-            raise ValueError(f'ERROR: Unknown trade type "{user_action["action_type"]}".')
+            raise ValueError(f'ERROR: Unknown trade type "{agent_action["action_type"]}".')
```

Comment:
I don't think this syntax will work because the `UserAction` dataclass does not have the required dunder methods.